Backend/Remora.Discord.Gateway/BaseGatewayClient.cs

```diff
@@ -0,0 +1,572 @@
+//
+//  BaseGatewayClient.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Remora.Discord.API;
+using Remora.Discord.API.Abstractions.Gateway;
+using Remora.Discord.API.Abstractions.Gateway.Commands;
+using Remora.Discord.API.Abstractions.Gateway.Events;
+using Remora.Discord.Gateway.Results;
+using Remora.Discord.Gateway.Services;
+using Remora.Discord.Gateway.Transport;
+using Remora.Results;
+
+namespace Remora.Discord.Gateway;
+
+/// <summary>
+/// Provides an abstract implementation of an <see cref="IGatewayClient"/>
+/// that must be extended with relevant connection and resumption flows.
+/// </summary>
+public abstract class BaseGatewayClient : IGatewayClient, IAsyncDisposable
+{
+    private readonly ILogger<BaseGatewayClient> _logger;
+    private readonly Channel<IPayload> _payloadSendQueue;
+    private readonly Dictionary<Guid, IGatewayCommand> _preShutdownCommands;
+
+    private Task<Result>? _sendTask;
+    private Task<Result>? _receiveTask;
+    private Task<Result>? _responderDispatchTask;
+    private volatile bool _isRunning;
```

Comment:
I think it may be better to use a normal bool here and use `Volatile.Read`/`Volatile.Write` instead

Reply:
Aye. I was thinking of a video I saw about the wonky behavior of `volatile` but looking at what Fred has said about it I don't think it matters here. 
![D3542525-6037-440F-BC55-F35227604D8B](https://user-images.githubusercontent.com/42438262/160310950-c4950ef4-026a-4a65-aa92-f50f5ec5cf33.jpeg)